### PR TITLE
#191: Threadsafe rendering

### PR DESCRIPTION
**Pull request by espen42**

### CODE REVIEW

**Review by rymsha — COMMENTED** [review]
This is unclear to me how objects are borrowed from a pool. Where is the pool?

**Review by espen42 — COMMENTED** [review]
The pool and the borrow/return logic is in ServerSideRenderer.java (which I noticed was hidden in the Files Changed view because of the diff size). I marked them up here for clarity:

https://github.com/enonic/lib-react4xp/pull/205/files#diff-c43fd25de8e3ec04538707aa4eae1fb9ad7002b65c8943e2d1484411fd1ef852R73

File: src/main/java/com/enonic/lib/react4xp/ssr/ServerSideRenderer.java [hunk 1]
```diff
@@ -30,291 +21,136 @@
...
-            );
+            LOG.error(new ErrorHandler().getLoggableStackTrace(e, null));
+            result = Map.of(ErrorHandler.KEY_ERROR, e.getMessage());
         }
-    }
 
-    @Override
-    public void initialize(BeanContext context) {
-        this.RESOURCE_SERVICE_SUPPLIER = context.getService(ResourceService.class);
+        if (renderer != null) {
```
Comment (rymsha) [inline src/main/java/com/enonic/lib/react4xp/ssr/ServerSideRenderer.java:146]: should be in `finally` block

File: src/main/java/com/enonic/lib/react4xp/ssr/ServerSideRenderer.java [hunk 2]
```diff
@@ -30,291 +21,136 @@
...
-                    getSolutionTips());
-
-            if (IS_PRODMODE && assetsInvolved != null) {
-                for (String asset : assetsInvolved) {
-                    CacheMarker cacheMarker = ASSET_CACHE_MARKERS.get(asset);
-                    synchronized (cacheMarker) {
-                        cacheMarker.isCached = false;
-                    }
-                }
+            if (renderer != null) {
```
Comment (rymsha) [inline src/main/java/com/enonic/lib/react4xp/ssr/ServerSideRenderer.java:121]: should be in `finally` block

Comment (espen42) [reply src/main/java/com/enonic/lib/react4xp/ssr/ServerSideRenderer.java]: Done - this and the other one.

Out of curiosity: this is after a `catch`, which should catch _all exceptions_ and just log them before proceeding - should be no `return` or other flow changes inside the catch block, so you'd expect the `if (renderer!=null) {` etc will run no matter what, with or without a finally block - _unless an error occurs during the catch block itself._ 

Is this - errors my happen during the catch but the finally will run no matter what - the only reason for using a finally block here, or are there other best-practice reasons too, even in this case?

**Review by rymsha — APPROVED** [review]

**Pull request merged**